File: java/src/main/java/org/zero123/PyScriptEngine/ModSdk/World.java
package org.zero123.PyScriptEngine.ModSdk;

import net.minecraft.core.BlockPos;
import org.zero123.PyScriptEngine.Utils.BlockUtil;
import org.zero123.PyScriptEngine.Utils.WorldUtil;

public class World
{
    // 服务端设置方块
    public static int _serverSetBlock(String pos, String blockJo, int oldHandling, int dimensionId, int updateNeighbors)
    {
        var posArray = Utils.parseIntArray(pos);
        final var levelOpt = WorldUtil.findServerLevelById(dimensionId);
        if(levelOpt.isEmpty())
        {
            return 0;
        }
        final var level = levelOpt.get();   // Level对象
        final var blockState = BlockUtil.jsonToBlockState(blockJo);
        final var blockPos = new BlockPos(posArray[0], posArray[1], posArray[2]);
        final var current = level.getBlockState(blockPos);  // 获取已有方块对象
        switch (oldHandling)
        {
            case 2:
                if (!current.isAir())
                {
                    return 0; // 保留旧方块
                }
                break;
            case 1:
                if (!current.isAir())
                {
                    if (!level.isClientSide())
                    {
                        level.destroyBlock(blockPos, true); // 掉落
                    }
                }
                break;
            case 0:
            default:
                // 直接替换不作判定
                break;
        }
        int flags = 2 + (updateNeighbors != 0 ? 1 : 0);
        boolean setState = level.setBlock(blockPos, blockState, flags);
        return (setState ? 1 : 0);
    }

    // 服务端获取方块信息
    public static String _serverGetBlock(String pos, int dimensionId)
    {
        var posArray = Utils.parseIntArray(pos);
        final var levelOpt = WorldUtil.findServerLevelById(dimensionId);
        if(levelOpt.isEmpty())
        {
            return "{\"name\":\"minecraft:air\", \"aux\": 0}";
        }
        final var level = levelOpt.get();   // Level对象
        final var blockPos = new BlockPos(posArray[0], posArray[1], posArray[2]);
        final var current = level.getBlockState(blockPos);  // 获取已有方块对象
        return BlockUtil.blockStateToJson(current).toString();
    }
}

File: python/mod/api/world.py
from .jni import findJavaCls, CAST_TYPE, numberTupleJoin
from json import dumps, loads
lambda: "By Zero123"

_WORLD_MODULE = "org/zero123/PyScriptEngine/ModSdk/World"

def _serverSetBlock(pos: tuple[int, int, int], blockJo: dict, oldHandling: bool = 0, dimensionId: int = 0, updateNeighbors: bool=True) -> bool:
    """ 服务端设置方块
    :param pos: 方块位置
    :param blockJo: 方块数据 {"name": "minecraft:stone"} (AUX在此废弃，不会被解析)
    :param oldHandling: 旧方块处理策略: 0(默认).直接替换 1.破坏并替换 2.保留
    :param dimensionId: 维度ID
    :param updateNeighbors: 是否更新邻居方块(默认更新)
    """
    return bool(findJavaCls(_WORLD_MODULE, "_serverSetBlock", [CAST_TYPE.STRING, CAST_TYPE.STRING, CAST_TYPE.INT, CAST_TYPE.INT, CAST_TYPE.INT], CAST_TYPE.INT).call(
        numberTupleJoin(int(v) for v in pos), dumps(blockJo) if blockJo else "", int(oldHandling), int(dimensionId), int(updateNeighbors)
    ))

def _serverGetBlock(pos: tuple[int, int, int], dimensionId: int = 0) -> dict:
    """ 服务端获取方块数据
    :param pos: 方块位置
    :param dimensionId: 维度ID
    :return: 方块数据字典 {"name": "minecraft:stone", "aux": 0}
    """
    return loads(findJavaCls(_WORLD_MODULE, "_serverGetBlock", [CAST_TYPE.STRING, CAST_TYPE.INT], CAST_TYPE.STRING).call(
        numberTupleJoin(int(v) for v in pos), int(dimensionId)
    ))